### DISCUSSION TITLE
\htmlId and subscripts / superscripts

### DISCUSSION THREAD

**Discussion by ashok-khanna** [KaTeX/KaTeX#3675] [General] [answered]
If I try `x_\htmlId{myId}{3}` (e.g. in katex.org/#demo) I get the following error: KaTeX parse error: Got function '\htmlId' with no arguments as subscript at position 3: x_\̲h̲t̲m̲l̲I̲d̲{myId}{3}

Similar issue with `^`. However this will work `x_{\htmlId{myId}{3}}`. Should I just use that or is there a way to fix the above? Also, does {} do anything in particular - couldn't find a link in the docs that describe it precisely. I assumed it was a grouping operator.

This is an awesome project, thank you!!!

**Comment by JacobBeaudry** [comment, marked as answer]
Hi Ashok, 

That's the expected behaviour. As you guessed, the braces are basically groupers; they have various effects on the final output.

In general, KaTeX seeks to recreate LaTeX's behaviour, and LaTeX requires nested subscripts to be wrapped in braces, e.g. 
   - `x_1_i` 
  
will cause an error, but 
  - `x_{1_i}` 
 
 will not. The same holds for any command that takes an argument, placed in a subscript, so 
  - `x_\sin{x}` 

will give an error, because of `\sin`'s argument, however, 
  - `x_\oplus` 

will not, because `\oplus` does not take an argument. 

In your case, the `\htmlId{}{}` command takes two arguments, so the error you're getting is the expected behaviour. 

In general, until you're very comfortable with LaTeX, you can avoid a lot of confusing, hard-to-understand bugs if you always include the braces. 

If you're new to TeX / LaTeX _etc._ I'd suggest taking a peek at the [tex.se](https://tex.stackexchange.com/) and [overleaf](https://www.overleaf.com/learn/latex/Tutorials) sites. 

At the former, you'll find detailed answers to a broad range of questions from some of the most knowledgeable members of the TeX community (just be sure to read the rules of the site before posting); at the latter, you'll find well-documented introductions and tutorials. 

All the best,

Jacob

**Comment by ashok-khanna** [comment]
Thank you very much Jacob for the useful answer.
